scripts/prepare_release.py

```diff
@@ -0,0 +1,873 @@
+                      rf'\g<1>{version}\g<2>{date}',
+                      old_content, flags=re.MULTILINE, count=1)
+
+    def run(self) -> None:
+        """Assemble the changelog if needed."""
+        subprocess.check_call(['framework/scripts/assemble_changelog.py'],
+                              cwd=self.info.top_dir)
+        if self.files_are_clean('ChangeLog') and \
+           self.info.old_version != self.info.version:
+            # Edge case: no change since the previous release.
+            # This could happen, for example, in an emergency release
+            # of Mbed TLS to ship a crypto bug fix, or when testing
+            # the release script.
+            self.edit_file('ChangeLog', self.create_section)
+        else:
+            self.edit_file('ChangeLog', self.finalize_release)
+        self.git_commit_maybe(['ChangeLog', 'ChangeLog.d'],
+                              'Assemble changelog and set release date')
+
+
+class BumpVersionStep(Step):
+    """Bump the product version and commit the result.
+
+    Do nothing if the product version is already as expected.
+
+    Note: this step does not check or affect submodules.
+
+    Note: this step does not currently handle ABI version bumps,
+    only product version bumps.
+    """
+
+    @classmethod
+    def name(cls) -> str:
+        return 'version'
+
+    # Files and directories that may contain version information.
+    FILES_WITH_VERSION = [
+        'CMakeLists.txt',
+        'doxygen',
+        'include',
+        'tests/suites',
+    ]
+
+    def run(self) -> None:
+        """Bump the product version if needed."""
+        subprocess.check_call(['scripts/bump_version.sh',
+                               '--version', self.info.version],
+                              cwd=self.info.top_dir)
+        self.git_commit_maybe(self.FILES_WITH_VERSION,
+                              'Bump version to ' + self.info.version)
+
+
+class ArchiveStep(Step):
+    """Prepare release archives and the associated checksum file."""
+
+    @classmethod
+    def name(cls) -> str:
+        return 'archive'
+
+    def tar_git_files(self, plain_tar_path: str, prefix: str) -> None:
+        """Create an uncompressed tar files with the git contents."""
+        # We archive the index, not the commit 'HEAD', because
+        # we may have modified some files to turn off GEN_FILES.
+        # We do use the commit mtime rather than the current time,
+        # however, for reproducibility.
+        # A downside of not archiving a commit is that the archive won't
+        # contain an extended header with the commit ID that
+        # `git get-tar-commit-id` could retrieve. If we change to releasing
+        # an exact commit, we should make sure that the commit gets published.
+        index = self.git_index_as_tree_ish()
+        mtime = self.commit_timestamp()
+        self.call_git(['archive', '--format=tar',
+                       '--mtime', str(mtime),
```

Comment:
`git archive --mtime` requires a recent git. Ubuntu 22.04 isn't good enough. Ubuntu 24.04 is ok.

That's a reasonable assumption for a developer machine. Unfortunately, our CI doesn't have an Ubuntu 24.04 image yet, only 22.04.

We could avoid this by making a commit with the `GEN_FILES` change.

Reply:
Instead of `git archive --mtime=…`, which can't be used in our CI yet, we could use `faketime`, which is already present in our docker images.

Reply:
We need to update the Dockerfile anyway, to set the user name and email so that Git can make a commit. And I have a working Ubuntu 24.04 Docker image with little fuss. So I think the way to go is to add the Ubuntu 24.04 image, which is long overdue anyway.